### ISSUE TITLE
Some SSB queries are hang on 99% of execution

### ISSUE THREAD

**Issue by rudolfninja** [ClickHouse/ClickHouse#69196]
**Company or project name**

ClickHouse

**Describe what's wrong**

When running slightly modified SSB queries 4.1-4.3 it seems that the query execution is hang pn 99%.

**Does it reproduce on the most recent release?**

Yes

**How to reproduce**

1. Prepare tables data and create tables according to scenario described in [Star Schema Benchmark](https://clickhouse.com/docs/en/getting-started/example-datasets/star-schema) scenario with following differences: a) use `-s 100` parameted of dbgen to generate data; b) do not create `lineorder_flat` table;
2. Run following query (it is slightly modified query 4.1):
`SELECT C_NATION, SUM(LO_REVENUE - LO_SUPPLYCOST) AS profit FROM customer, supplier, part, lineorder WHERE LO_CUSTKEY = C_CUSTKEY AND LO_SUPPKEY = S_SUPPKEY AND LO_PARTKEY = P_PARTKEY AND C_REGION = 'AMERICA' AND S_REGION = 'AMERICA' AND (P_MFGR = 'MFGR#1' OR P_MFGR = 'MFGR#2') GROUP BY C_NATION ORDER BY C_NATION SETTINGS join_algorithm = 'auto';`
=> Query execution will hang on 99% (see attached screenshot)

**Expected behavior**

Query execution result.

**Error message and/or stacktrace**

![Screenshot 2024-09-03 155630](https://github.com/user-attachments/assets/ebe03efb-d389-48ec-ba5a-0084757ae941)

**Additional context**

It doesn't matter if experemental analyzer is set or not.
Any execptions or errors are absent in server's log

**Comment by rudolfninja** [comment]
May be the same problem as described in https://github.com/ClickHouse/ClickHouse/issues/40602

**Comment by vdimir** [comment]
It happens because of a lack of optimization to rewrite a comma join to a hash join. If optimization is not applied, the algorithm falls back to an ineffective cross join. You can check this by setting `cross_to_inner_join_rewrite = 2`, which will generate an exception:

```
DB::Exception: Failed to rewrite ', ...' to INNER JOIN: no equi-join conditions found in WHERE clause. You may set the setting `cross_to_inner_join_rewrite` to `1` to allow a slow CROSS JOIN for this case. (INCORRECT_QUERY)
(query: SELECT C_NATION, SUM(LO_REVENUE - LO_SUPPLYCOST) AS profit FROM customer, supplier, part, lineorder)
```

However, we can help the query choose the correct algorithm by changing the order of tables to `SELECT C_NATION, SUM(LO_REVENUE - LO_SUPPLYCOST) AS profit FROM lineorder, customer, supplier, part`. In that case, ClickHouse is able to determine the correct keys.

Of course, the optimizer should be smarter and able to perform such optimization without query rewrite for any order of tables. This is planned, but there is no ETA yet.

Thanks for reporting this!

**Comment by rudolfninja** [comment]
> It happens because of a lack of optimization to rewrite a comma join to a hash join. If optimization is not applied, the algorithm falls back to an ineffective cross join. You can check this by setting `cross_to_inner_join_rewrite = 2`, which will generate an exception:
> 
> ```
> DB::Exception: Failed to rewrite ', ...' to INNER JOIN: no equi-join conditions found in WHERE clause. You may set the setting `cross_to_inner_join_rewrite` to `1` to allow a slow CROSS JOIN for this case. (INCORRECT_QUERY)
> (query: SELECT C_NATION, SUM(LO_REVENUE - LO_SUPPLYCOST) AS profit FROM customer, supplier, part, lineorder)
> ```
> 
> However, we can help the query choose the correct algorithm by changing the order of tables to `SELECT C_NATION, SUM(LO_REVENUE - LO_SUPPLYCOST) AS profit FROM lineorder, customer, supplier, part`. In that case, ClickHouse is able to determine the correct keys.
> 
> Of course, the optimizer should be smarter and able to perform such optimization without query rewrite for any order of tables. This is planned, but there is no ETA yet.
> 
> Thanks for reporting this!

I checked your suggestion and indeed it works. Could you please spcify which class makes decision about optimization and perform optimization? I'll try to check it by myslef and may be implement quick solution for my case.

**Comment by vdimir** [comment]
> Could you please spcify which class makes decision about optimization and perform optimization?

It's  https://github.com/ClickHouse/ClickHouse/blob/master/src/Analyzer/Passes/CrossToInnerJoinPass.cpp 
Let me know if you have any further question regarding the code

**Comment by vdimir** [comment]
Yes, a QueryTree is essentially an AST (Abstract Syntax Tree) combined with contextual information, such as references from columns to source tables, types of expressions, and more. Several passes are executed on top of the query tree, usually involving simple syntactic optimizations and query rewrites. However, more complex optimizations are typically performed on the query plan, which is built from the query tree at a later stage. You can find more details in the ClickHouse source code [here](https://github.com/ClickHouse/ClickHouse/blob/master/src/Processors/QueryPlan/Optimizations/Optimizations.h).

You can check how passes modify the query tree using:

```sql
EXPLAIN QUERY TREE dump_passes = 1, dump_ast = 1, passes = <N> SELECT ...
```

where `<N>` is the number of passes to run.

Rewriting a cross join to an inner join is currently implemented as a separate pass. This pass analyzes the `WHERE` clause for a particular pair of tables to extract conditions that can be used for a hash join. However, this level of optimization lacks flexibility; for example, reordering tables can be quite tricky. We are considering moving such optimizations to the plan level, although this is also not a trivial task.
